crates/bevy_render/src/mesh/primitives/extrusion.rs

```diff
@@ -0,0 +1,380 @@
+use bevy_math::{
+    primitives::{Annulus, Capsule2d, Circle, Ellipse, Extrusion, Primitive2d},
+    Vec2, Vec3,
+};
+
+use crate::mesh::{Indices, Mesh, VertexAttributeValues};
+
+use super::{MeshBuilder, Meshable};
+
+/// A type representing a segment of the perimeter of an extrudable mesh.
+pub enum PerimeterSegment {
+    /// This segment of the perimeter will be shaded smooth.
+    ///
+    /// This has the effect of rendering the segment's faces with softened edges, so it is appropriate for curved shapes.
+    ///
+    /// The normals for the vertices that are part of this segment will be calculated based on the positions of their neighbours.
+    /// Each normal is interpolated between the normals of the two line segments connecting it with its neighbours.
+    /// Closer vertices have a stronger effect on the normal than more distant ones.
+    ///
+    /// Since the vertices corresponding to the first and last indices do not have two neighbouring vertices, their normals must be provided manually.
+    Smooth {
+        /// The normal of the first vertex.
+        first_normal: Vec2,
+        /// The normal of the last vertex.
+        last_normal: Vec2,
+        /// A list of indices representing this segment of the perimeter of the mesh.
+        ///
+        /// The indices must be ordered such that the *outside* of the mesh is to the right
+        /// when walking along the vertices of the mesh in the order provided by the indices.
+        ///
+        /// For geometry to be rendered, you must provide at least two indices.
+        indices: Vec<u32>,
+    },
+    /// This segment of the perimeter will be shaded flat.
+    ///
+    /// This has the effect of rendering the segment's faces with hard edges.
+    Flat {
+        /// A list of indices representing this segment of the perimeter of the mesh.
+        ///
+        /// The indices must be ordered such that the *outside* of the mesh is to the right
+        /// when walking along the vertices of the mesh in the order provided by indices.
+        ///
+        /// For geometry to be rendered, you must provide at least two indices.
+        indices: Vec<u32>,
+    },
+}
+
+impl PerimeterSegment {
+    fn vertices_per_layer(&self) -> usize {
+        match self {
+            PerimeterSegment::Smooth { indices, .. } => indices.len(),
+            PerimeterSegment::Flat { indices } => 2 * (indices.len() - 1),
+        }
+    }
+    fn indices_per_segment(&self) -> usize {
```

Comment:
Some documentation on these would be nice; it wasn't exactly clear to me what "indices per segment" should mean